File: algo_cpp/bj2150.cpp
#include <cstdio>
#include <algorithm>
#include <vector>
using namespace std;

vector<int> adj[10001];
vector<int> stack;
vector<vector<int>> ans;
int dfsn[10001],sccn[10001];
bool finished[10001];
int sn = 0, cnt = 0;

int dfs(int u){
    dfsn[u] = ++cnt;
    stack.push_back(u);
    
    int res = dfsn[u];
    for(auto v : adj[u]){
        if(dfsn[v]==0){
            int tmp = dfs(v);
            res = res<tmp?res:tmp;
        }else if(!finished[v]){
            res = res<dfsn[v]?res:dfsn[v];
        }
    }
    if(res==dfsn[u]){
        vector<int> tmpScc;
        while(1){
            int tmp = stack.back();stack.pop_back();
            tmpScc.push_back(tmp);
            finished[tmp] = 1;
            sccn[tmp] = sn;
            if(tmp==u)
              break;
        }
        sort(tmpScc.begin(),tmpScc.end());
        ans.push_back(tmpScc);
        sn++;
    }
    return res;
}

int main(void){
    int v,e;
    
    scanf("%d %d",&v,&e);
    for(int i = 0; i < e; i++){
        int u,v;
        scanf("%d %d",&u,&v);
        adj[u].push_back(v);
    }
    for(int i = 1; i <= v; i++){
        if(dfsn[i]==0)
            dfs(i);
    }
    
    sort(ans.begin(),ans.end());
    printf("%d\n",sn);
    for(auto scc : ans){
        for(auto tmp : scc)
            printf("%d ",tmp);
        printf("-1\n");
    }
    return 0;
}

File: algo_py/boj/bj2150.py
import sys
sys.setrecursionlimit(10**9)
input = sys.stdin.readline
sn = cnt = 0

def dfs(u):
    global sn
    global cnt
    cnt+=1
    dfsn[u] = cnt
    stack.append(u)

    res = dfsn[u]
    for v in adj[u]:
        if dfsn[v]==0:
            res = min(res,dfs(v))
        elif not finished[v]:
            res = min(res,dfsn[v])
    
    if res == dfsn[u]:
        tmpScc = []
        while 1:
            tmp = stack.pop()
            tmpScc.append(tmp)
            finished[tmp] = True
            sccn[tmp] = sn
            if tmp==u:
                break
        tmpScc.sort()
        ans.append(tmpScc)
        sn+=1
    return res


if __name__ == '__main__':
    V, E = map(int, input().split())
    adj = [[] for _ in range(V+1)]
    dfsn = [0 for _ in range(V+1)]
    sccn = [0 for _ in range(V+1)]
    finished = [False for _ in range(V+1)]
    ans = []
    stack = []
    for _ in range(E):
        u,v = map(int,input().split())
        adj[u].append(v)
    for i in range(1,V+1):
        if dfsn[i]==0:
            dfs(i)
    ans.sort()
    print(sn)
    for scc in ans:
        sys.stdout.write(' '.join(map(str,scc))+' -1\n')